File: Validbrackets.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        n_brackets = len(s)
        if n_brackets == 0:
            return True
        for i in range(n_brackets):
            char = s[i]
            if len(stack) == 0:
                stack.append(char)
            else:
                match_flag = self.is_match(stack[-1], char)
                if match_flag:
                    stack.pop(-1)
                else:
                    stack.append(char)

        return len(stack) == 0

    def is_match(self, a, b):
        if a == '(':
            return b == ')'
        if a == '[':
            return b == ']'
        if a == '{':
            return b == '}'

File: Validbrackets.cpp
//
// Created by Golfer on 2020/7/21.
//

#include <string>
#include <stack>
using namespace std;

class Solution {
public:
    bool isValid(string s) {
        stack<char> my_stack;
        int n = s.size();
        if (n == 0)
            return true;
        for (char & c : s) {
            if (my_stack.empty())
                my_stack.push(c);
            else {
                bool match_flag = is_match(my_stack.top(), c);
                if (match_flag)
                    my_stack.pop();
                else
                    my_stack.push(c);
            }
        }
        return my_stack.empty();
    }

    bool is_match(char a, char b) {
        if (a == '(')
            return b == ')';
        if (a == '[')
            return b == ']';
        if (a == '{')
            return b == '}';
        return false;
    }
};

int main() {
    string s = "()";
    Solution solution = Solution();
    bool valid = solution.isValid(s);
    printf("%d", valid);
}

